src/combobox/ComboBox.ts

```diff
@@ -266,118 +267,158 @@ export default class ComboBox extends ComboBoxBase<ComboBoxProperties> {
-			this._restoreFocus();
+			if (this._callInputFocus) {
+				this._callInputFocus = false;
+				const input = element.querySelector('input') as HTMLElement;
+				input && input.focus();
+			}
 		}
 	}
 
-	protected onElementUpdated(element: HTMLElement, key: string) {
-		if (key === 'root') {
-			this._restoreFocus();
-			const selectedResult = element.querySelector('[data-selected="true"]');
-			selectedResult && this._scrollIntoView(<HTMLElement> selectedResult);
-		}
+	protected renderInput(): WNode {
+		const {
+			clearable,
+			disabled,
+			inputProperties = {},
+			invalid,
+			readOnly,
+			required,
+			results = [],
+			value = '',
+			theme
+		} = this.properties;
+
+		return w(TextInput, <TextInputProperties> {
+			...inputProperties,
+			key: 'textinput',
+			'aria-activedescendant': this._getResultId(results[this._activeIndex], this._activeIndex),
+			'aria-owns': this._getMenuId(),
+			classes: this.classes(clearable ? css.clearable : null),
+			controls: this._getMenuId(),
+			disabled,
+			invalid,
+			onBlur: this._onInputBlur,
+			onFocus: this._onInputFocus,
+			onInput: this._onInput,
+			onKeyDown: this._onInputKeyDown,
+			readOnly,
+			required,
+			theme,
+			value
+		});
 	}
 
-	protected renderMenu(results: any[]): WNode | null {
-		const { theme = {}, isResultDisabled, CustomResultMenu = ResultMenu, CustomResultItem } = this.properties;
+	protected renderClearButton() {
+		const {
+			disabled,
+			readOnly
+		} = this.properties;
+
+		return v('button', {
+			'aria-controls': this._getMenuId(),
+			classes: this.classes(css.clear),
+			disabled,
+			readOnly,
+			onclick: this._onClearClick
+		}, [
+			'clear combo box',
+			v('i', { classes: this.classes(iconCss.icon, iconCss.closeIcon),
+				role: 'presentation', 'aria-hidden': 'true'
+			})
+		]);
+	}
+
+	protected renderMenuButton() {
+		const {
+			disabled,
+			readOnly
+		} = this.properties;
+
+		return v('button', {
+			classes: this.classes(css.trigger),
+			disabled,
+			readOnly,
+			tabIndex: -1,
+			onclick: this._onArrowClick
+		}, [
+			'open combo box',
+			v('i', {
+				'aria-hidden': 'true',
+				classes: this.classes(iconCss.icon, iconCss.downIcon),
+				role: 'presentation'
+			})
+		]);
+	}
+
+	protected renderMenu(results: any[]): DNode | null {
+		const { theme, isResultDisabled } = this.properties;
 
 		if (results.length === 0 || !this._open) {
 			return null;
 		}
 
-		return w(CustomResultMenu, {
-			getResultLabel: this._getResultLabel,
-			CustomResultItem,
-			id: this._menuId,
-			isResultDisabled,
-			onResultMouseDown: this._onResultMouseDown,
-			onResultMouseEnter: this._onResultMouseEnter,
-			onResultMouseUp: this._onResultMouseUp,
-			results,
-			selectedIndex: this._activeIndex,
-			theme
-		});
+		return v('div', {
+			key: 'dropdown',
+			classes: this.classes(css.dropdown),
+			onmouseover: this._onResultHover,
+			onmousedown: this._onResultMouseDown
+		}, [
+			w(Listbox, {
```

Comment:
Did we lose the ability to pass custom result items and menus into the ComboBox?

Reply:
Yup, I think we talked about not passing around sub-widget constructors anymore, right? I think Ant or Matt G originally wanted them gone entirely, but it makes handling events more difficult. I suppose ComboBox/Select could take an extension of Listbox though, since it's a standalone widget.

Reply:
I think that the ability to easily customize items and the menu itself is important. I'm not tied to passing in a custom class, but we need some way to accomplish the same functionality that was provided previously. I don't think the APIs should change drastically because of the Listbox change.

Reply:
I could definitely add the `CustomMenu` that takes an extension of Listbox back in. It seemed (at least to me) that we were moving in the direction of having most or all customization happen through extension. I'm definitely open to other ideas, though.

Reply:
I can get on board with extension and no sub-widgets as long as the offloaded render methods are extremely small and simple to override.